### DISCUSSION TITLE
Startup time

### DISCUSSION THREAD

**Discussion by lars-reimann** [mwaskom/seaborn#3677] [Q&A] [answered]
Simply importing `seaborn` is quite slow, even on a fast computer:

```
(Measure-Command { python -c "import seaborn" }).TotalSeconds
```

➡️ 1.2129588

```
(Measure-Command { python -c "import matplotlib" }).TotalSeconds
```

➡️ 0.1960905

I've used seaborn 0.13.2, matplotlib 3.8.3, and Python 3.12.2 for this quick comparison.

Running ` python -X importtime -c "import seaborn"` reveals that a few transitively imported packages constitute most of the delay, e.g.:

* `seaborn` > `seaborn.widgets` > `ipywidgets`: 0.3s
* `seaborn.relational` > `seaborn._statistics` > `scipy.stats`: 0.3s

While ugly, using lazy imports for those would already cut the startup time in half, with little effort. The first invocation of an API element that needs those imports would be slower, of course, but if they are never used, there's no cost at all.

Maybe a library like [`apipkg`](https://pypi.org/project/apipkg/) would also help with lazy-importing packages. Together with `TYPE_CHECKING` imports in `__init__` files, IDEs can still figure out what gets imported, e.g. for one of our projects:

```py
import apipkg

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ._column import Column

apipkg.initpkg(__name__, {
    "Column": "._column:Column",
})

__all__ = [
    "Column",
]
```

Are lazy imports, at least for a few packages, an option for you at all, or do you want to load everything upfront?

**Comment by mwaskom** [comment]
👎 on a new dependency

that said, the widget stuff can probably just get deprecated, I am not sure anyone uses it.

but matplotlib and pandas and numpy are huge; I am not sure I buy that the marginal import time of seaborn itself is that significant.

**Reply by lars-reimann** [reply]
> but matplotlib and pandas and numpy are huge; I am not sure I buy that the marginal import time of seaborn itself is that significant.

`python -X importtime -c "import seaborn"` shows where exactly time is spent for importing. The overhead from seaborn itself is very small indeed. The startup time rather stems from the fact that seaborn's top-level `__init__.py` file imports all modules of seaborn, which eagerly import their dependencies.

An `import seaborn`, thus also imports, say, `ipywidgets` (taking 0.3s) to power the widgets that according to you nobody uses. Here, moving the imports into the few functions that need them would help (or deprecating widgets altogether).

**Comment by thuiop** [comment]
Are you sure of your results ? The `ipywidgets` import is about 100µs on my computer, and while` scipy.stats` is indeed longer, it is on the same order as `pandas` for instance, so I don't know if the benefit would be huge.

**Reply by lars-reimann** [reply]
Make sure to check the import time in an environment that actually has `ipywidgets`. I've tried it on two devices and got consistent results.

**Reply by thuiop** [reply]
Ah, right, that was it.

**Comment by mwaskom** [comment, marked as answer]
Also `ipywidgets` and `scipy` are both optional dependencies so if you're doing something like building a container environment where fast startup is important, you could just not include them.

**Reply by lars-reimann** [reply]
That helped improve the situation:

In the environment I was testing this, we have `jupyter` as a dependency for the docs. This also depends on `ipywidgets`, which was then picked up and imported by `seaborn`, increasing the import time. Still not ideal that `ipywidgets` is eagerly imported in this case, but we are no longer affected by that. Thanks.

**Comment by Doggie52** [comment]
I'm taking somewhere between 6-15 seconds to import seaborn on my end. It seems very random, sometimes it's blazingly fast. I don't have ipywidgets installed. What could I do to help troubleshooting where this extreme slowness is coming from?
